### ISSUE TITLE
[WAGON-405] Preemptive authentication still not working with Maven 3.0.4+

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-wagon#462]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** opened **[WAGON-405](https://issues.apache.org/jira/browse/WAGON-405?redirect=false)** and commented

Although by default the preemptive authentication should work now with Maven 3.0.4, a wireshare dump exposes, that each request for downloading a new dependency does not initially come with the authentication header. Only after  Nexus replied with a 401 the right credentials are set in the authentication header. According to [1], since Maven 3.0.4 the default HTTP wagon uses preemptive authentication which cannot be disabled [2]. Even with Maven 3.1.1 it does not work.

[1] - http://maven.apache.org/guides/mini/guide-http-settings.html#Maven_3.0.4
[2] - http://maven.40175.n5.nabble.com/Maven-3-0-4-preemptive-auth-problem-tt5470892.html#none

---

**Attachments:**
- [WAGON-405-MVN3.1.1.log](https://issues.apache.org/jira/secure/attachment/12720705/WAGON-405-MVN3.1.1.log) (_99.36 kB_)

**Issue Links:**
- [MNG-4792](https://issues.apache.org/jira/browse/MNG-4792) Preemptive authentication doesn't work

3 votes, 8 watchers

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

My previous observation is not correct. Indeed preemptive authentication could work, as this is set here: https://github.com/apache/maven-wagon/blob/wagon-2.4/wagon-providers/wagon-http-shared4/src/main/java/org/apache/maven/wagon/shared/http4/AbstractHttpClientWagon.java#L762, but the default value for usePreemptive is still false (https://github.com/apache/maven-wagon/blob/wagon-2.4/wagon-providers/wagon-http-shared4/src/main/java/org/apache/maven/wagon/shared/http4/HttpMethodConfiguration.java#L55) and it is unclear how to overwrite that value for regular GET methods.

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Finally figured out that enabling preemptive authentication works with

```
<servers>
 <server>
  <id>serverid</id>
  <username>myuser</username>
  <password>mypassword</password>
  <configuration>
   <wagonProvider>httpclient</wagonProvider>
   <httpConfiguration>
    <all>
     <usePreemptive>true</usePreemptive>
    </all>
   </httpConfiguration>
  </configuration>
 </server>
</servers>
```

Please update the documentation at http://maven.apache.org/guides/mini/guide-http-settings.html#Example:_Using_Preemptive_Authentication and fix the statement that preemptive authentication is enabled by default or make it really the default within the code.

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Even with the new pull request merged (https://github.com/apache/maven-wagon/pull/6) by default there is no preemptive authentication being used.

**Comment by jira-importer** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

won't be "by default"!

**Comment by jira-importer** [comment]
**[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin)** commented

Ok, but with the most recent source code I see the following behaviour:

1) Preemptive authentication used always for PUT (https://github.com/apache/maven-wagon/blob/master/wagon-providers/wagon-http/src/main/java/org/apache/maven/wagon/providers/http/AbstractHttpClientWagon.java#L526)
2) Preemptive authentication only used for GET if explicitly set in the config (https://github.com/apache/maven-wagon/blob/master/wagon-providers/wagon-http/src/main/java/org/apache/maven/wagon/providers/http/AbstractHttpClientWagon.java#L711). Never used by default.

To me that doesn't make much sense to me. Please make the documentation in http://maven.apache.org/guides/mini/guide-http-settings.html#Maven_3.0.4 clearer with regard to what should be the default and also please add an example how to enable preemptive authentication in http://maven.apache.org/guides/mini/guide-http-settings.html#Example:_Using_Preemptive_Authentication. Thanks a lot.

**Comment by jira-importer** [comment]
**[Mirko Friedenhagen](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=mfriedenhagen)** commented

[Konrad Windszus](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kwin), my guess:
* On central AKA repo1.maven.org, anyone is allowed to download stuff, so GET defaults to using no authentication.
* On the other hand, for uploading, PUTting new stuff, you do not want this to be done anonymously.

I am with you, that the documentation should be updated.

**Comment by jira-importer** [comment]
**[Christian Kibsgaard Lunde](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christian.lunde@vegvesen.no)** commented

I also just wasted a whole day trying to figure out why the example in the documentation for preemptive authentication doesn't work, The documentation really needs to get updated!
